skcosmo/pcovr/pcovr.py

```diff
@@ -0,0 +1,313 @@
+        self.whiten = False
+        self._eig_solver = partial(
+            eig_solver, n_components=self.n_components, tol=self.tol, add_null=True
+        )
+
+    def fit(self, X, Y, Yhat=None, W=None):
+        """
+
+        Fit the model with X and Y. Depending on the dimensions of X,
+        calls either `_fit_feature_space` or `_fit_structure_space`
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            Training data, where n_samples is the number of samples and
+            n_features is the number of features.
+        Y : array-like, shape (n_samples, n_properties)
+            Training data, where n_samples is the number of samples and
+            n_properties is the number of properties
+        Yhat : array-like, shape (n_samples, n_properties), optional
+            Regressed training data, where n_samples is the number of samples and
+            n_properties is the number of properties. If not supplied, computed
+            by ridge regression.
+        W : array-like, shape (n_features, n_properties), optional
+            Weights of regressed training data. If not supplied, computed
+            by ridge regression.
+
+        """
+
+        X, Y = check_X_y(X, Y, y_numeric=True, multi_output=True)
+
+        if self.space is not None and self.space not in [
+            "feature",
+            "structure",
+            "auto",
+        ]:
+            raise ValueError("Only feature and structure space are supported.")
+
+        if self.n_components is None:
+            self.n_components = min(X.shape)
+
+        if Yhat is None or W is None:
+            Yhat, W = self._compute_Yhat(X, Y, Yhat=Yhat, W=W)
+
+        if self.space is None:
+            if X.shape[0] > X.shape[1]:
+                self.space = "feature"
+            else:
+                self.space = "structure"
+
+        if self.space == "feature":
+            self._fit_feature_space(X, Yhat, W)
+        else:
+            self._fit_structure_space(X, Yhat, W)
+
+        self.mean_ = np.mean(X, axis=0)
+        self.pxy_ = self.pxt_ @ self.pty_
+        if len(Y.shape) == 1:
+            self.pxy_ = self.pxy_.reshape(
+                X.shape[1],
+            )
+            self.pty_ = self.pty_.reshape(
+                self.n_components,
+            )
+
+        self.components_ = self.pxt_.T  # for sklearn compatibility
+
+    def _compute_Yhat(self, X, Y, Yhat=None, W=None):
+        """
+        Method for computing the approximation of Y to fit the PCovR
+        """
+
+        if Yhat is None:
+            if W is None:
+                lr = LR(**self.lr_args)  # some sort of args
+                lr.fit(X, Y)
+                Yhat = lr.predict(X)
+                W = lr.coef_.T
+            else:
+                Yhat = X @ W
+
+        elif W is None:
+            W = np.linalg.pinv(np.dot(X.T, X), rcond=self.regularization)
+            W = np.linalg.multi_dot([W, X.T, Y])
```

Comment:
The computation of `W` is probably better as a `np.linalg.lstsq` or similar, which is more stable. We should try to avoid inverses and pseudoinverses wherever possible.